core/arch/aarch64/emit_utils.c

```diff
@@ -750,11 +750,12 @@ emit_indirect_branch_lookup(dcontext_t *dc, generated_code_t *code, byte *pc,
     /* x1 now holds the fragment_entry_t* in the hashtable. */
     APP(&ilist, load_tag);
     /* Load tag from fragment_entry_t* in the hashtable to x0. */
-    /* ldr x0, [x1, #tag_fragment_offset] */
+    /* ldr x0, [x1], #tag_fragment_offset */
     APP(&ilist,
-        INSTR_CREATE_ldr(
+        INSTR_CREATE_ldr_imm(
             dc, opnd_create_reg(DR_REG_X0),
-            OPND_CREATE_MEMPTR(DR_REG_X1, offsetof(fragment_entry_t, tag_fragment))));
+            opnd_set_zero_offset_post_index(OPND_CREATE_MEMPTR(DR_REG_X1, 0)),
```

Comment:
This isn't making sense to me: the immediate/displacement for the writeback is not zero but is `sizeof(fragment_entry_t)`, and the separate immediate operand has to match the memory operand displacement.  I would think that for a non-zero immediate, the encoder will figure out whether it's a pre-indexed or post-indexed add by looking at the memory operand to see whether the displacement there is zero: if it's zero, it's post-indexed; if it's non-zero, it's pre-indexed.  The flag `pre_index` inside the memory operand is only needed if you have a 0 immediate (i.e., you're adding 0 to the base reg) and the encoder can't tell which is which and will pick the post-indexed and you really care about the actual raw bits of the encoding so you want the pre-indexed encoding (both are semantically the same).

We need help from the ARM folks for their encoder: @AssadHashmi if you could clarify: is that how the AArch64 encoder works to figure out pre vs post indexed?  IIRC that's what the ARM encoder does.  So this instruction should *not* need to set any special flags on the memory operand (and a flag on the memory operand should not be affecting whether there's a 2nd destination, the base reg).

Reply:
Yes, you shouldn't need to do anything special to the memory operand, the decoder/encoder take care of such things, e.g. see `mem9post` mask in `codec.txt`
Is there any reason why `opnd_create_base_disp_aarch64()` is not being used here?

Reply:
Yes, I had noticed the existence of the opnd_create_base_disp_aarch64() and it's used only once by decoder and is used by a couple of unit tests. I probably misunderstood the design of codec a bit. Thank you for the comments, I will make a change to let the encoder take care of the pre-/post-indexed selection.